### PR TITLE
Remove some duplicata of broadcast txn from ChannelMonitor

### PR DESCRIPTION
**Pull request by ariard**
First part of refactoring ChannelMonitor, removing duplicata transaction for single-signed transactions. Doing it separately from actual ChannelMonitor to ease review.

Removing duplicata avoid some pitfalls

- locking twice bumping inputs if we have to CPFP some transactions (not-yet-implemented)
- transaction origin inferences if tx broadcaster doesn't dedup tx itself
- useless signing operations
- undefined behavior with an external signer implementing some signature replay protections

A second refactoring should move remaining dual-signed transactions, to do latter or while implementing option_simplified_commitment.

### CODE REVIEW

File: lightning/src/ln/channelmonitor.rs [hunk 1]
```diff
@@ -1772,11 +1772,11 @@ impl ChannelMonitor {
 					let mut inputs_info = Vec::new();
 
 					macro_rules! sign_input {
-						($sighash_parts: expr, $input: expr, $amount: expr, $preimage: expr) => {
+						($sighash_parts: expr, $input: expr, $amount: expr, $preimage: expr, $idx: expr) => {
```
Comment (TheBlueMatt) [inline lightning/src/ln/channelmonitor.rs]: If you're gonna stop relying on the stupid hack of shoving the idx in the sequence field (which is a good cleanup), can you do it in its own commit and fully remove it instead of only partially?

Comment (ariard) [reply lightning/src/ln/channelmonitor.rs]: That was a quick fix, given than in #462, the macro and a bunch of codes around is going away. Split in its own commit anyway (bundled there because of the very basic dedup parser in test_utils relying on nSequence)

Comment (TheBlueMatt) [reply lightning/src/ln/channelmonitor.rs]: Right, but its easier to review if you just do it all the way...this patch seems to pass tests:

```
diff --git a/lightning/src/ln/channelmonitor.rs b/lightning/src/ln/channelmonitor.rs
index d55db634..8a393be9 100644
--- a/lightning/src/ln/channelmonitor.rs
+++ b/lightning/src/ln/channelmonitor.rs
@@ -1805,17 +1805,16 @@ impl ChannelMonitor {
                                                        }
                                                        if let Some(payment_preimage) = self.payment_preimages.get(&htlc.payment_hash) {
                                                                if htlc.offered {
-                                                                       let mut input = TxIn {
+                                                                       let input = TxIn {
                                                                                previous_output: BitcoinOutPoint {
                                                                                        txid: commitment_txid,
                                                                                        vout: transaction_output_index,
                                                                                },
                                                                                script_sig: Script::new(),
-                                                                               sequence: idx as u32, // reset to 0xfffffffd before sign_input
+                                                                               sequence: 0xff_ff_ff_fd,
                                                                                witness: Vec::new(),
                                                                        };
                                                                        if htlc.cltv_expiry > height + CLTV_SHARED_CLAIM_BUFFER {
-                                                                               input.sequence = 0xff_ff_ff_fd;
                                                                                inputs.push(input);
                                                                                inputs_desc.push(if htlc.offered { InputDescriptors::OfferedHTLC } else { InputDescriptors::ReceivedHTLC });
                                                                                inputs_info.push((payment_preimage, tx.output[transaction_output_index as usize].value, htlc.cltv_expiry, idx));
@@ -1834,8 +1833,6 @@ impl ChannelMonitor {
                                                                                let height_timer = Self::get_height_timer(height, htlc.cltv_expiry);
                                                                                let mut used_feerate;
                                                                                if subtract_high_prio_fee!(self, fee_estimator, single_htlc_tx.output[0].value, predicted_weight, used_feerate) {
-                                                                                       let idx = single_htlc_tx.input[0].sequence;
-                                                                                       single_htlc_tx.input[0].sequence = 0xff_ff_ff_fd;
                                                                                        let sighash_parts = bip143::SighashComponents::new(&single_htlc_tx);
                                                                                        let (redeemscript, htlc_key) = sign_input!(sighash_parts, single_htlc_tx.input[0], htlc.amount_msat / 1000, payment_preimage.0.to_vec(), idx);
                                                                                        assert!(predicted_weight >= single_htlc_tx.get_weight());
@@ -1868,7 +1865,7 @@ impl ChannelMonitor {
                                                                                vout: transaction_output_index,
                                                                        },
                                                                        script_sig: Script::new(),
-                                                                       sequence: idx as u32,
+                                                                       sequence: 0xff_ff_ff_fd,
                                                                        witness: Vec::new(),
                                                                };
                                                                let mut timeout_tx = Transaction {
@@ -1884,8 +1881,6 @@ impl ChannelMonitor {
                                                                let height_timer = Self::get_height_timer(height, htlc.cltv_expiry);
                                                                let mut used_feerate;
                                                                if subtract_high_prio_fee!(self, fee_estimator, timeout_tx.output[0].value, predicted_weight, used_feerate) {
-                                                                       let idx = timeout_tx.input[0].sequence;
-                                                                       timeout_tx.input[0].sequence = 0xff_ff_ff_fd;
                                                                        let sighash_parts = bip143::SighashComponents::new(&timeout_tx);
                                                                        let (redeemscript, htlc_key) = sign_input!(sighash_parts, timeout_tx.input[0], htlc.amount_msat / 1000, vec![0], idx);
                                                                        assert!(predicted_weight >= timeout_tx.get_weight());

```

**Comment by TheBlueMatt** [comment]
Still needs rebase, I think.

**Comment by ariard** [comment]
Is it ready or are you expecting https://github.com/rust-bitcoin/rust-lightning/pull/461#discussion_r371368274 sorry ?

**Comment by TheBlueMatt** [comment]
I'd prefer to see the that patch included just cause it makes review much more obvious.

**Comment by ariard** [comment]
Somehow took your patch

File: lightning/src/ln/functional_tests.rs [hunk 2]
```diff
@@ -2175,15 +2167,25 @@ fn claim_htlc_outputs_single_tx() {
...
-
-		assert_eq!(node_txn[0], node_txn[7]);
-		assert_eq!(node_txn[1], node_txn[8]);
-		assert_eq!(node_txn[2], node_txn[9]);
-		assert_eq!(node_txn[3], node_txn[10]);
-		assert_eq!(node_txn[4], node_txn[11]);
-		assert_eq!(node_txn[3], node_txn[5]); //local commitment tx + htlc timeout tx broadcasted by ChannelManger
+		assert_eq!(node_txn.len(), 26); // ChannelManager : 2, ChannelMontitor: 8 (1 standard revoked output, 2 revocation htlc tx, 1 local commitment tx + 1 htlc timeout tx) * 2 (block-rescan) + 5 * (1 local commitment tx + 1 htlc timeout tx)
+
+		// 3 = 5 = 7 = 9 = 14 = 17 = 20 = 23
```
Comment (TheBlueMatt) [inline lightning/src/ln/functional_tests.rs]: Whats the intuition between these all being the same (can you add a comment for it?)? The comment doesn't seem to indicate they should be.

File: lightning/src/ln/functional_tests.rs [hunk 3]
```diff
@@ -2314,14 +2316,16 @@ fn test_htlc_on_chain_success() {
 		_ => panic!("Unexpected event"),
 	};
 	macro_rules! check_tx_local_broadcast {
-		($node: expr, $htlc_offered: expr, $commitment_tx: expr, $chan_tx: expr) => { {
-			// ChannelManager : 3 (commitment tx, 2*HTLC-Timeout tx), ChannelMonitor : 2 (timeout tx) * 2 (block-rescan)
+		($node: expr, $htlc_offered: expr, $commitment_tx: expr, $chan_tx: expr, $len: expr) => { {
+			// ChannelManager : 3 (commitment tx, 2*HTLC-Timeout tx), ChannelMonitor : 2 (timeout tx)
```
Comment (TheBlueMatt) [inline lightning/src/ln/functional_tests.rs]: This comment is similarly out of date (can you explain why one case has 7 and one 5) and can you just use the $htlc_offered flag to check for 7 vs 5 instead of adding a new parameter?

File: lightning/src/ln/functional_tests.rs [hunk 4]
```diff
@@ -2502,8 +2506,7 @@ fn test_htlc_on_chain_timeout() {
 	nodes[0].block_notifier.block_connected(&Block { header, txdata: vec![commitment_tx[0].clone()]}, 200);
 	check_closed_broadcast!(nodes[0], false);
 	let node_txn = nodes[0].tx_broadcaster.txn_broadcasted.lock().unwrap().clone(); // ChannelManager : 2 (commitment tx, HTLC-Timeout tx), ChannelMonitor : 2 (timeout tx) * 2 block-rescan
-	assert_eq!(node_txn.len(), 4);
-	assert_eq!(node_txn[0], node_txn[3]);
+	assert_eq!(node_txn.len(), 3);
```
Comment (TheBlueMatt) [inline lightning/src/ln/functional_tests.rs]: Comment above is out of date (just needs to lose the "*2" I think).

File: lightning/src/ln/functional_tests.rs [hunk 5]
```diff
@@ -4147,14 +4148,14 @@ fn test_onchain_to_onchain_claim() {
 	let commitment_tx = nodes[0].node.channel_state.lock().unwrap().by_id.get_mut(&chan_1.2).unwrap().channel_monitor().get_latest_local_commitment_txn();
 	nodes[1].block_notifier.block_connected(&Block { header, txdata: vec![commitment_tx[0].clone()]}, 1);
 	let b_txn = nodes[1].tx_broadcaster.txn_broadcasted.lock().unwrap();
-	assert_eq!(b_txn.len(), 4);
-	check_spends!(b_txn[1], chan_1.3); // Local commitment tx, issued by ChannelManager
-	check_spends!(b_txn[2], b_txn[1]); // HTLC-Success tx, as a part of the local txn rebroadcast by ChannelManager in the force close
-	assert_eq!(b_txn[0], b_txn[3]); // HTLC-Success tx, issued by ChannelMonitor, * 2 due to block rescan
+	// ChannelMonitor: claim tx, ChannelManager: local commitment tx
```
Comment (TheBlueMatt) [inline lightning/src/ln/functional_tests.rs]: This says 2, but then we check for 3?

File: lightning/src/ln/functional_tests.rs (hunk 2 above)
Comment (ariard) [reply lightning/src/ln/functional_tests.rs]: It's for local commitment tx + HTLC-Timeout tx duplicata, right now they are still duplicated (even as of #466, planning to do it in its own PR because transaction generation of this class of txn isn't actually a thing in as-of-master `bump_claim_tx`

File: lightning/src/ln/functional_tests.rs [hunk 6]
```diff
@@ -2175,15 +2167,25 @@ fn claim_htlc_outputs_single_tx() {
...
 		let node_txn = nodes[1].tx_broadcaster.txn_broadcasted.lock().unwrap();
-		assert_eq!(node_txn.len(), 29); // ChannelManager : 2, ChannelMontitor: 8 (1 standard revoked output, 2 revocation htlc tx, 1 local commitment tx + 1 htlc timeout tx) * 2 (block-rescan) + 5 * (1 local commitment tx + 1 htlc timeout tx)
-
-		assert_eq!(node_txn[0], node_txn[7]);
-		assert_eq!(node_txn[1], node_txn[8]);
-		assert_eq!(node_txn[2], node_txn[9]);
-		assert_eq!(node_txn[3], node_txn[10]);
-		assert_eq!(node_txn[4], node_txn[11]);
-		assert_eq!(node_txn[3], node_txn[5]); //local commitment tx + htlc timeout tx broadcasted by ChannelManger
+		assert_eq!(node_txn.len(), 26); // ChannelManager : 2, ChannelMontitor: 8 (1 standard revoked output, 2 revocation htlc tx, 1 local commitment tx + 1 htlc timeout tx) * 2 (block-rescan) + 5 * (1 local commitment tx + 1 htlc timeout tx)
```
[the remark this answers is not in the document]
Comment (ariard) [reply lightning/src/ln/functional_tests.rs]: Updated with better comment, see log commented if you want to check : https://github.com/ariard/rust-lightning/commit/6fa4bd0b8415973c6a146932ce5bb64517daf104

File: lightning/src/ln/functional_tests.rs (hunk 3 above)
Comment (ariard) [reply lightning/src/ln/functional_tests.rs]: There is diff between node 0 and node 1 because in one case the commitment tx is local and in the other case that's remote so same explanation than comment above (local HTLC-timeout are still duplicated)

File: lightning/src/ln/functional_tests.rs (hunk 5 above)
Comment (ariard) [reply lightning/src/ln/functional_tests.rs]: Oops, check is false, comment right, deleted too much. Should be fix

**Comment by ariard** [comment]
Updated at 2668817 with better comments.

>  At some point this tx broadcast checking mess is going to need to be cleaned up.

Best we can do is to have a check-template for every type of transaction and pass the expected list of templates to test method on tx_broadcaster. And find a way to desecribe tx graph dependency in an easy way

File: lightning/src/ln/functional_tests.rs [hunk 7]
```diff
@@ -2314,14 +2319,17 @@ fn test_htlc_on_chain_success() {
 		_ => panic!("Unexpected event"),
 	};
 	macro_rules! check_tx_local_broadcast {
-		($node: expr, $htlc_offered: expr, $commitment_tx: expr, $chan_tx: expr) => { {
-			// ChannelManager : 3 (commitment tx, 2*HTLC-Timeout tx), ChannelMonitor : 2 (timeout tx) * 2 (block-rescan)
+		($node: expr, $htlc_offered: expr, $commitment_tx: expr, $chan_tx: expr, $len: expr) => { {
```
Comment (TheBlueMatt) [inline lightning/src/ln/functional_tests.rs]: The added $len parameter is still here (and unused).

File: lightning/src/ln/functional_tests.rs [hunk 8]
```diff
@@ -2455,10 +2462,11 @@ fn test_htlc_on_chain_timeout() {
 	let timeout_tx;
 	{
 		let mut node_txn = nodes[1].tx_broadcaster.txn_broadcasted.lock().unwrap();
-		assert_eq!(node_txn.len(), 8); // ChannelManager : 2 (commitment tx, HTLC-Timeout tx), ChannelMonitor : 6 (HTLC-Timeout tx, commitment tx, timeout tx) * 2 (block-rescan)
-		assert_eq!(node_txn[0], node_txn[5]);
-		assert_eq!(node_txn[1], node_txn[6]);
-		assert_eq!(node_txn[2], node_txn[7]);
+		assert_eq!(node_txn.len(), 7); // ChannelManager : 2 (commitment tx, HTLC-Timeout tx), ChannelMonitor : 6 (HTLC-Timeout tx, commitment tx, timeout tx)
```
Comment (TheBlueMatt) [inline lightning/src/ln/functional_tests.rs]: The comment is still wrong here.

File: lightning/src/ln/functional_tests.rs [hunk 9]
```diff
@@ -6801,9 +6806,7 @@ fn test_bump_penalty_txn_on_remote_commitment() {
 	let feerate_preimage;
 	{
 		let mut node_txn = nodes[1].tx_broadcaster.txn_broadcasted.lock().unwrap();
-		assert_eq!(node_txn.len(), 7); // 2 * claim tx (broadcasted from ChannelMonitor) * 2 (block-reparsing) + local commitment tx + local HTLC-timeout + HTLC-success (broadcasted from ChannelManager)
-		assert_eq!(node_txn[0], node_txn[5]);
-		assert_eq!(node_txn[1], node_txn[6]);
+		assert_eq!(node_txn.len(), 5); // 2 * claim tx (broadcasted from ChannelMonitor) + local commitment tx + local HTLC-timeout (broadcasted from ChannelManager)
```
Comment (TheBlueMatt) [inline lightning/src/ln/functional_tests.rs]: This comment looks wrong?

**Comment by ariard** [comment]
494219e, more comments fixed
